File: src/4/4.cpp
#include <iostream>
#include <sys/resource.h>
#include <unistd.h>
#include <vector>
#include <string>

using namespace std;

void printMemoryUsage(const string &note) {
    struct rusage usage;
    getrusage(RUSAGE_SELF, &usage);
    // ru_maxrss is in kilobytes on Linux
    cout << "[" << note << "] Memory usage: " << usage.ru_maxrss / 1024.0 << " MB" << endl;
}

int main() {
    printMemoryUsage("Start");
    
    // Allocate a vector with 1,000,000 integers
    vector<int> v;
    for (int i = 0; i < 1000000; i++) {
        v.push_back(i);
    }
    printMemoryUsage("After allocation");
    
    // Clear the vector to free memory
    v.clear();
    printMemoryUsage("After deletion");
    
    return 0;
}


// Output
// [Start] Memory usage: 2.03906 MB
// [After allocation] Memory usage: 7.59375 MB
// [After deletion] Memory usage: 7.59375 MB

File: src/4/4.rs
// Stub version for environments without external crate support.
fn print_memory_usage(note: &str) {
    println!("[{}] (Memory usage info unavailable in this environment)", note);
}

fn main() {
    print_memory_usage("Start");
    
    // Allocate a vector with 1,000,000 integers
    let v: Vec<i32> = (0..1_000_000).collect();
    print_memory_usage("After allocation");
    
    // Drop the vector and observe memory usage
    drop(v);
    print_memory_usage("After deletion");
}

/*
use sysinfo::{System, SystemExt};

fn print_memory_usage(note: &str) {
    let mut sys = System::new_all();
    sys.refresh_all();
    let used_memory = sys.used_memory() as f64 / 1024.0 / 1024.0; // Convert kilobytes to MB
    println!("[{}] Memory usage: {:.2} MB", note, used_memory);
}

fn main() {
    print_memory_usage("Start");
    
    // Allocate a vector with 1,000,000 integers
    let v: Vec<i32> = (0..1_000_000).collect();
    print_memory_usage("After allocation");
    
    // Drop the vector and observe memory usage
    drop(v);
    print_memory_usage("After deletion");
}
*/
